src/utils.jl

```diff
@@ -45,7 +45,7 @@ macro generate_interface(expr)
     
     calculator_type = nothing
     try 
-        calculator_type = get_calculator_type(expr)
+        calculator_type = get_calculator_type(expr, type)
     catch _
```

Comment:
I would suggest to be more specific here. Catch all can lead to very hard to debug error suppressions.

Reply:
are you asking to specify which exception to catch? I don't have too strong a view here but I guess the message could be more specific, i.e. it should state what test fails and then print the exception so one can see during debugging what went wrong?

Reply:
Yes I'm worried about the exception catching.

Reply:
@tjjarvinen  or  @CedricTravelletti  -- not sure which of you wrote this, but can you please push a fix?

Reply:
I think you are now missing the context here. What that code actually does is

```julia
try 
        calculator_type = get_calculator_type(expr, type)
    catch _
        error("Could not determine calculators type")
end
```

Or basically it wraps and error with better description on what is happening, so that **it is easier to debug**. Rather that throwing a error message that user does not understand comes from.

Reply:
I dont think anybody is missing context. IIUC Michaels concern is that not all exceptions mean that the method isn't implemented.

Reply:
The function that is called `get_calculator_type` does

```julia
function get_calculator_type(expr, type)
    if type[:type] in [:Energy, :Forces, :Virial]   # Calculator interface (low-level)
        return expr.args[1].args[end - 2].args[2]
    else
        return expr.args[1].args[end].args[2]   # high-level interface
    end
end
```

That can result in at least `BoundsError`, `UndefVarError` and `MethodError`. All thease mean that the calculator function definition that the user made is not accordingly to the definition of AtomsCalculators API. So, we basically want to catch all errors on that point and tell the user that they are submitting incorrect definition.

In part this error is meant to be part of the testing, to say users that you are using the interface incorrecly.

We could of course specify it more on the individual errors, by telling how the user is implementing the interface incorrecly. But, that would make this code much more complicated. So, to me it is just simpler to tell the user that your implementation is incorrect, go and look for the documentation again.

Reply:
@mfherbst  ---  can you please comment. Personally it would be enough for me to just give a better error message, e.g. to say explicitly exactly what failed. This is something that cannot currently be seen from "Could not determine calculators type". But maybe you had more in mind.

Reply:
I think that is a good start. My main worry is that someone makes a typo in the implementation of their calculator thereby triggering an exception at macro evaluation. Instead of  standard error message they see this and line numbers and stack trace now point at the wrong spot. If that's not the case I'm ok.

Reply:
I improved the error message. It now suggest that there is a typo in function definition related to the calculator type.

This should be now clear enough error
```julia
try 
        calculator_type = get_calculator_type(expr, type)
    catch _
        error("Possible typo (or other issue) in function definition. Could not determine the calculators type.")
    end
```

Reply:
@tjjarvinen  -  thanks for trying to fix this. I'm not convinced though that this addresses @mfherbst  concern.  Maybe he can comment.  Maybe what could be done is to do `@error` instead of `error` and then rethrow the exception. That way the developer can see the stacktrace and see what went wrong.

Reply:
I think it's a lot better now. I've taken the liberty to merge this to not have this drag on too long. I've opened #23.